audit/base/base.go

```diff
@@ -0,0 +1,556 @@
+	// will be requeued here. When this is full, the record is dropped.
+	QueueSize int
+}
+
+// defaults sets the default values for the settings. It returns a copy of settings to avoid an allocation.
+func (s Settings) defaults() Settings {
+	if s.QueueSize < 1 {
+		s.QueueSize = MaxQueueSize
+	}
+	return s
+}
+
+// Option represents an option for the audit client.
+type Option func(c *Client) error
+
+// WithSettings sets the settings for the audit client. If not set, the default values are used.
+func WithSettings(s Settings) Option {
+	return func(c *Client) error {
+		c.settings = s.defaults()
+		return nil
+	}
+}
+
+// WithLogger sets the logger for the audit client. If not set, the default logger is used.
+// This is for the few log messages that can occur, such as message drops. As a note, slog.Logger
+// is not an interface. But loggers are created with a handler which tell the logger where to
+// send the log messages. You can use that to convert a slog.Logger to a different logger if needed.
+// Also, check with your logging package, as it may have a way to send slog.Logger messages to it.
+func WithLogger(l *slog.Logger) Option {
+	return func(c *Client) error {
+		c.log = l
+		return nil
+	}
+}
+
+// New returns a new audit client.
+func New(c conn.Audit, options ...Option) (client *Client, err error) {
+	if c == nil {
+		return nil, Error{Err: fmt.Errorf("cannot pass a conn.Audit that is nil"), Category: ErrConnection}
+	}
+
+	cli := &Client{
+		settings:          Settings{}.defaults(),
+		log:               slog.Default(),
+		heartbeatInterval: 30 * time.Minute,
+		done:              make(chan struct{}),
+	}
+	cli.conn.Store(&c)
+
+	defer func() {
+		if err != nil {
+			c.CloseSend(ctxBack)
+		}
+	}()
+
+	for _, o := range options {
+		if err := o(cli); err != nil {
+			return nil, err
+		}
+	}
+
+	kver, err := kernelVer()
+	if err != nil {
+		return nil, fmt.Errorf("could not determine kernel version on platform: %w", err)
+	}
+
+	cli.heartbeat = msgs.Msg{
+		Type: msgs.Heartbeat,
+		Heartbeat: msgs.HeartbeatMsg{
+			AuditVersion: version.Semantic,
+			OsVersion:    kver,
+			Language:     runtime.Version(),
+			Destination:  c.Type().String(),
+		},
+	}
+	// This handles a special case where the version of the language is not goX.X.X, but instead
+	// the git hash of the compiler. This happens on custom builds of the go compiler.
+	if !strings.HasPrefix(cli.heartbeat.Heartbeat.Language, "go") {
+		cli.heartbeat.Heartbeat.Language = "go" + cli.heartbeat.Heartbeat.Language
+	}
+
+	cli.sendCh = make(chan streamData[SendMsg], cli.settings.QueueSize)
+	cli.stopSend = make(chan chan struct{}, 1)
+
+	go cli.sender()
+
+	return cli, nil
+}
+
+// Conn returns the current connection to the audit server. The one inside
+// Client can be changed out from under you. This is provided for testing purposes.
+// This is not supported for users.
+func (c *Client) Conn() conn.Audit {
```

Comment:
If its only for testing purposes why not just leave the test in the same package and use `c.conn` without exposing this at all?

Reply:
Because it is used in tests for the next package you are reviewing.